src/common/transformations/src/transformations/common_optimizations/matmul_split_decomposition.cpp

```diff
@@ -0,0 +1,209 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "transformations/common_optimizations/matmul_split_decomposition.hpp"
+
+#include "itt.hpp"
+#include "openvino/core/rt_info.hpp"
+#include "openvino/opsets/opset1.hpp"
+#include "openvino/pass/pattern/op/or.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+
+using namespace ov;
+using namespace ov::pass::pattern;
+
+bool pass::MatmulGatherDecomposition::split_weights(const Output<Node>& weights,
+                                                    OutputVector& new_weights,
+                                                    Output<Node>* bias,
+                                                    OutputVector& new_bias,
+                                                    const bool transpose_b) {
+    // weights is static
+    if (weights.get_partial_shape().size() != 2u) {
+        return false;
+    }
+
+    if (bias) {
+        const auto& bias_rank = bias->get_partial_shape().size();
+        if (bias_rank != 3 && bias_rank != 1) {
+            return false;
+        }
+    }
+
+    // Decompose weights
+    auto axis = register_new_node(op::v0::Constant::create(element::i32, Shape{}, {transpose_b ? 0 : 1}));
+    auto split = register_new_node<op::v1::Split>(weights, axis, decompose_num);
+    for (auto& out : split->outputs()) {
+        new_weights.emplace_back(out);
+    }
+
+    if (bias) {
+        // Decompose bias
+        auto axis2 = register_new_node(op::v0::Constant::create(element::i32, Shape{}, {-1}));  // axis -1
+        auto split2 = register_new_node<op::v1::Split>(*bias, axis2, decompose_num);
+        for (auto& out : split2->outputs()) {
+            new_bias.emplace_back(out);
+        }
+    }
+    return true;
+}
+
+pass::MatmulGatherDecomposition::MatmulGatherDecomposition() {
```

Comment:
I understand that this transformation is trying to match a very specific pattern from llm models but shouldn't we have some heuristic for the weights size or something.
I mean do we expect any model with any weights sizes to benefit from this transformation?
Also, please describe in the commit message / PR description the motivation of having this transformation, why we expect it to speed up llms in the first place.

Reply:
Yes, I want to match VIT similar structure model, and I also add some heuristics, check [Rank](https://github.com/xipingyan/openvino/blob/4d5f03d96a2f747aea887aaa7d39bd3507d36272/src/common/transformations/src/transformations/common_optimizations/matmul_split_decomposition.cpp#L62), [decompose_num](https://github.com/xipingyan/openvino/blob/4d5f03d96a2f747aea887aaa7d39bd3507d36272/src/common/transformations/src/transformations/common_optimizations/matmul_split_decomposition.cpp#L67), and specific [transpose order](https://github.com/xipingyan/openvino/blob/4d5f03d96a2f747aea887aaa7d39bd3507d36272/src/common/transformations/src/transformations/common_optimizations/matmul_split_decomposition.cpp#L93), do you think these are not enough?

Reply:
Probably this will be enough most of the times.
But, again, this is mostly about the reason we are getting the speed-ups.
I assume we observe speed-ups not because of the ranks, decompose_num and transpose order, but because we become less memory bound. But maybe I am wrong.

Reply:
Yes, I think it is related to input data size, because it is dynamic shape, so it is hard to custom describe it, just try to best.